### PR TITLE
improves table class; adds table deserializers; adds table deserialization test;

### PR DESCRIPTION
**Pull request by olgamelnichuk**

### CODE REVIEW

File: src/main/kotlin/ac/uk/ebi/biostd/serialization/json/AttributeJsonSerializer.kt [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        gen.writeStringField(NAME, attr.name)
+        gen.writeStringField(VALUE, attr.value)
+
+        if (attr.reference) {
+            gen.writeBooleanField(REFERENCE, attr.reference)
+        }
+
+        if (attr.terms.isNotEmpty()) {
+            gen.writeArrayFieldStart(TERMS)
+            attr.terms.forEach { gen.writeObject(it) }
```
Comment (Juan-EBI) [inline src/main/kotlin/ac/uk/ebi/biostd/serialization/json/AttributeJsonSerializer.kt]:
```
attr.terms.forEach { gen::writeObject }
```

File: src/main/kotlin/ac/uk/ebi/biostd/serialization/json/TableJsonSerializer.kt [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    private val tableDeserializer = object : TableJsonDeserializer<Section>(Section::class.java) {
+        override fun createTable(elements: List<Section>): Table<Section> {
+            return SectionsTable(elements)
+        }
+    }
+
+    override fun deserialize(jp: JsonParser, ctxt: DeserializationContext): SectionsTable {
+        return tableDeserializer.deserialize(jp) as SectionsTable
+    }
+}
```
Comment (Juan-EBI) [inline src/main/kotlin/ac/uk/ebi/biostd/serialization/json/TableJsonSerializer.kt]:
```
companion : TableJsonDeserializer<Section>(Section::class.java) {
    override fun createTable(elements: List<Section>) = SectionsTable(elements)
}

???
```

File: src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt [hunk 3]
```diff
@@ -37,21 +41,54 @@ data class TableHeader(val name: String, val termNames: List<String> = listOf())
...
-})
+    fun headers(): List<TableHeader> = attributes.map { TableHeader(it.name, it.terms.names()) }
 
-class SectionsTable(type: String, parentAccNo: String) : Table<Section>("[$type][$parentAccNo]", {
-    TableRow(it.accNo, "accNo", it.attributes)
-})
+    fun values(headers: List<TableHeader>): List<String> =
+            headers.map { header -> findAttrByName(header.name) }
+                    .flatMap { attr -> listOf(attr.value) + attr.terms.values() }
+
```
Comment (Juan-EBI) [inline src/main/kotlin/ac/uk/ebi/biostd/common/Table.kt]:
```
headers.map { it -> findAttrByName(it.name) }. flatMap { attr -> listOf(attr.value) + attr.terms.values() }

File: src/test/kotlin/ac/uk/ebi/biostd/serialization/json/TableJsonSerializerTest.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        val table = JsonSerializer().deserialize(json, LinksTable::class.java)
+
+        val rows = table.elements
+        assertThat(rows.size).isEqualTo(2)
+        assertThat(rows[0].url).isEqualTo(url[0])
+        assertThat(rows[0].attributes.size).isEqualTo(2)
+        assertThat(rows[0].attributes[0]).isEqualTo(attr)
+
+        assertThat(rows[1].url).isEqualTo(url[1])
+        assertThat(rows[1].attributes.size).isEqualTo(0)
```
Comment (Juan-EBI) [inline src/test/kotlin/ac/uk/ebi/biostd/serialization/json/TableJsonSerializerTest.kt]: `assertThat(rows[1].attributes).hasSize(1)`

File: src/test/kotlin/ac/uk/ebi/biostd/serialization/json/TableJsonSerializerTest.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+            |   "url": "${url[1]}"
+            |}
+            |]""".trimMargin()
+
+        val table = JsonSerializer().deserialize(json, LinksTable::class.java)
+
+        val rows = table.elements
+        assertThat(rows.size).isEqualTo(2)
+        assertThat(rows[0].url).isEqualTo(url[0])
+        assertThat(rows[0].attributes.size).isEqualTo(2)
```
Comment (Juan-EBI) [inline src/test/kotlin/ac/uk/ebi/biostd/serialization/json/TableJsonSerializerTest.kt]: `assertThat(rows[0].attributes).hasSize(2)`

File: src/main/kotlin/ac/uk/ebi/biostd/serialization/json/AttributeJsonSerializer.kt (hunk 1 above)
Comment (olgamelnichuk) [reply src/main/kotlin/ac/uk/ebi/biostd/serialization/json/AttributeJsonSerializer.kt]: Done

File: src/main/kotlin/ac/uk/ebi/biostd/serialization/json/TableJsonSerializer.kt (hunk 2 above)
Comment (olgamelnichuk) [reply src/main/kotlin/ac/uk/ebi/biostd/serialization/json/TableJsonSerializer.kt]: with companion objects the code looks better to me

**Comment by Juan-EBI** [comment]
LGTM please merge

**Pull request merged**